src/ucp/dt/dt.c

```diff
@@ -143,37 +143,37 @@ size_t ucp_dt_pack(ucp_worker_h worker, ucp_datatype_t datatype,
     return result_len;
 }
 
-ucs_status_t ucp_dt_query(ucp_datatype_t datatype, const void *buffer, size_t count, ucp_datatype_attr_t *attr)
+ucs_status_t ucp_dt_query(ucp_datatype_t datatype, const void *buffer,
+                          size_t count, ucp_datatype_attr_t *attr)
 {
     ucp_dt_generic_t *dt_gen;
     void             *state_gen;
 
     /* Currently, the only datatype attribute to query is the packed size. */
     if (!(attr->field_mask & UCP_DATATYPE_ATTR_FIELD_PACKED_SIZE)) {
-        return UCS_ERR_UNSUPPORTED;
+        return UCS_OK;
     }
 
     switch (datatype & UCP_DATATYPE_CLASS_MASK) {
     case UCP_DATATYPE_CONTIG:
         attr->packed_size = ucp_contig_dt_elem_size(datatype);
         return UCS_OK;
     case UCP_DATATYPE_IOV:
-        if (buffer) {
-            attr->packed_size = ucp_dt_iov_length(buffer, count);
-            return UCS_OK;
+        if (buffer == NULL) {
+            return UCS_ERR_INVALID_PARAM;
         }
-        return UCS_ERR_INVALID_PARAM;
+        attr->packed_size = ucp_dt_iov_length(buffer, count);
+        return UCS_OK;
     case UCP_DATATYPE_GENERIC:
+        if (buffer == NULL) {
+            return UCS_ERR_INVALID_PARAM;
+        }
         dt_gen = ucp_dt_to_generic(datatype);
         ucs_assert(NULL != dt_gen);
-        if (buffer) {
-            state_gen = dt_gen->ops.start_pack(dt_gen->context, buffer, count);
-            attr->packed_size = dt_gen->ops.packed_size(state_gen);
-            return UCS_OK;
-        }
-        return UCS_ERR_INVALID_PARAM;
+        state_gen = dt_gen->ops.start_pack(dt_gen->context, buffer, count);
+        attr->packed_size = dt_gen->ops.packed_size(state_gen);
```

Comment:
need to also call dt_gen->ops.finish, or something may leak

Reply:
Added `dt_gen->ops.finish(state_gen);`